pkg/rtc/transportmanager.go

```diff
@@ -650,20 +658,45 @@ func (t *TransportManager) handleConnectionFailed(isShortLived bool) {
 		return
 	}
 
+	lowestPriorityConnectionType := t.getLowestPriorityConnectionType()
+
 	//
 	// Checking only `PreferenceSubscriber` field although any connection failure (PUBLISHER OR SUBSCRIBER) will
 	// flow through here.
 	//
 	// As both transports are switched to the same type on any failure, checking just subscriber should be fine.
 	//
 	getNext := func(ic *livekit.ICEConfig) livekit.ICECandidateType {
-		if ic.PreferenceSubscriber == livekit.ICECandidateType_ICT_NONE && t.params.ClientInfo.SupportsICETCP() && t.canUseICETCP() {
-			return livekit.ICECandidateType_ICT_TCP
-		} else if ic.PreferenceSubscriber != livekit.ICECandidateType_ICT_TLS && t.params.TURNSEnabled {
-			return livekit.ICECandidateType_ICT_TLS
-		} else {
-			return livekit.ICECandidateType_ICT_NONE
+		switch lowestPriorityConnectionType {
+		case types.ICEConnectionTypeUDP:
+			// try ICE/TCP if ICE/UDP failed
+			if ic.PreferenceSubscriber == livekit.ICECandidateType_ICT_NONE {
+				if t.params.ClientInfo.SupportsICETCP() && t.canUseICETCP() {
+					return livekit.ICECandidateType_ICT_TCP
+				} else if t.params.TURNSEnabled {
+					// fallback to TURN/TLS if TCP is not supported
+					return livekit.ICECandidateType_ICT_TLS
+				}
+			}
+
+		case types.ICEConnectionTypeTCP:
+			// try TURN/TLS if ICE/TCP failed,
+			// the configuration could have been ICT_NONE or ICT_TCP,
+			// in either case, fallback to TURN/TLS
+			if t.params.TURNSEnabled {
+				return livekit.ICECandidateType_ICT_TLS
+			} else {
+				// keep the current config
+				return ic.PreferenceSubscriber
```

Comment:
turn enabled could also mean UDP turn is enabled.. I think we need to pull out an explicit check for if TURN/TLS is enabled.

and if TCP failed when TLS is disabled, we should fall back to UDP? because it seems to invalidate the "UDP blocked" hypothesis

Reply:
oof.. my bad.. didn't spot the S hiding in there.. and thought this was the config options. thanks for clarifying